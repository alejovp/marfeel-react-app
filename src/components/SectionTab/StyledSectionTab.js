import styled from 'styled-components';

const getBorderBottom = ({ theme, isActive }) => {
    const { tipography, sectionMenu } = theme;
    const defaultColor = tipography && tipography.primaryFontColor;
    let borderColor;
    if (isActive && sectionMenu && sectionMenu.color) {
        borderColor = sectionMenu.color;
    } else if (isActive) {
        borderColor = defaultColor;
    } else {
        borderColor = 'transparent';
    }
    return `4px solid ${borderColor}`;
};

export const StyledSectionTab = styled.li`
    list-style: none;
    padding: 8px 10px;
    text-align: center;
    cursor: pointer;
    color: ${({ theme }) => theme.sectionMenu && theme.sectionMenu.color};
    border-bottom: ${getBorderBottom};

    &:hover {
        opacity: 0.8
    }
`;

StyledSectionTab.displayName = 'StyledSectionTab';
