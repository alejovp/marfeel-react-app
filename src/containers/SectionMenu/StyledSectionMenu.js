import styled from 'styled-components';

export const StyledSectionMenu = styled.ul`
    width: 100%;
    align-items: center;
    justify-content: flex-start;
    display: flex;
    margin-bottom: 0;
    padding: 0 ${({ theme }) => theme.header.padding}px;
    overflow-x: scroll;
    white-space: nowrap;
    font-size: ${({ theme }) => theme.sectionMenu && theme.sectionMenu.textSize || 1}em;
    transition: margin-top 0.4s;
    margin-top: ${({ hideSectionMenu, dimensions }) => hideSectionMenu ? -dimensions.height : 0}px;
    scrollbar-width: none;  /* Firefox */

    ::-webkit-scrollbar {
        display: none; /* Chrome, Safari and Opera */
    }
`;

StyledSectionMenu.displayName = 'StyledSectionMenu';
