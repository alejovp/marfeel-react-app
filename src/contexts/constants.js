export const SET_SECTION_MENU_HEIGHT = 'marfeel-react-app/header/SET_SECTION_MENU_HEIGHT';
