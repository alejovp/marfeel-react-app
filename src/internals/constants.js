export const APP_BAR_HEIGHT = 60;
export const SECTION_MENU_HEIGHT = 20;
export const BURGER_BUTTON_SIZE = 28;
export const FLOATING_CIRCLE_PADDING = 8;
