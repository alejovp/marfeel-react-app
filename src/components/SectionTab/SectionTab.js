import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { StyledSectionTab } from './StyledSectionTab';

export const SectionTab = ({ section = {}, onClick, currentSection }) => {

    const onClickHandler = useCallback(
        () => onClick(section.id), [onClick]
    );

    if (!section.id || !section.text) {
        return null;
    }

    const isActive = section.id === currentSection;

    return (
        <StyledSectionTab onClick={onClickHandler}
                          isActive={isActive}>
            <a>{section.text}</a>
        </StyledSectionTab>
    );
};

SectionTab.propTypes = {
    section: PropTypes.object.isRequired,
    onClick: PropTypes.func.isRequired,
    currentSection: PropTypes.string
};
