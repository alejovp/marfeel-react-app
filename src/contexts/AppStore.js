import React, { createContext, useReducer } from 'react';
import PropTypes from 'prop-types';
import { appReducer } from './reducer';
import { SECTION_MENU_HEIGHT } from '../internals/constants';
import { sectionsMock } from '../../__mocks__/sections';


const initialState = {
    sectionMenuHeight: SECTION_MENU_HEIGHT,
    openSideNav: false,
    currentSection: sectionsMock[0].id // this should be controlled by router location path
};

const StateContext = createContext();
const DispatchContext = createContext();

const AppStore = ({ children }) => {
    const [state, dispatch] = useReducer(appReducer, initialState);

    return (
        <DispatchContext.Provider value={dispatch}>
            <StateContext.Provider value={state}>
                {children}
            </StateContext.Provider>
        </DispatchContext.Provider>
    );
};

AppStore.propTypes = {
    children: PropTypes.node
};

export {
    AppStore,
    StateContext,
    DispatchContext
};
