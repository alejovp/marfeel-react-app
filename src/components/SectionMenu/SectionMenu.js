import React, { useCallback, useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { StyledSectionMenu } from './StyledSectionMenu';
import { SectionTab } from '../SectionTab';
import { useElemDimensions } from '../../hooks';
import { DispatchContext } from '../../contexts/AppStore';
import { setMenuSectionHeight } from '../../contexts/actions';


export const SectionMenu = ({ sections = [], hideSectionMenu }) => {
    const dispatch = useContext(DispatchContext);
    const [currentSection, setCurrentSection] = useState('sec-1');
    const [ref, dimensions] = useElemDimensions();
    const onClickHandler = useCallback(
        (section) => setCurrentSection(section), []
    );

    useEffect(() => {
        dispatch(setMenuSectionHeight(dimensions.height));
    }, [dimensions]);

    const renderSections = () => (
        sections.map(section => {
            return (
                <SectionTab key={section.id}
                            section={section}
                            currentSection={currentSection}
                            onClick={onClickHandler} />
            );
        })
    );

    return (
        <StyledSectionMenu ref={ref}
                           dimensions={dimensions}
                           hideSectionMenu={hideSectionMenu}>
            {renderSections()}
        </StyledSectionMenu>
    );
};

SectionMenu.propTypes = {
    sections: PropTypes.array.isRequired,
    hideSectionMenu: PropTypes.bool
};
