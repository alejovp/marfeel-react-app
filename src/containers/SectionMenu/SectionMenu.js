import React, { useCallback, useContext, useEffect } from 'react';
import PropTypes from 'prop-types';
import { StyledSectionMenu } from './StyledSectionMenu';
import { SectionTab } from '../../components/SectionTab';
import { useElemDimensions } from '../../hooks';
import { DispatchContext, StateContext } from '../../contexts/AppStore';
import { setCurrentSection, setMenuSectionHeight } from '../../contexts/actions';


export const SectionMenu = ({ sections = [], hideSectionMenu }) => {
    const dispatch = useContext(DispatchContext);
    const { currentSection } = useContext(StateContext);
    const [ref, dimensions] = useElemDimensions();
    const onClickHandler = useCallback(
        (section) => dispatch(setCurrentSection(section)), [dispatch]
    );

    useEffect(() => {
        dispatch(setMenuSectionHeight(dimensions.height));
    }, [dimensions]);

    const renderSections = () => (
        sections.map(section => {
            return (
                <SectionTab key={section.id}
                            section={section}
                            currentSection={currentSection}
                            onClick={onClickHandler} />
            );
        })
    );

    return (
        <StyledSectionMenu ref={ref}
                           dimensions={dimensions}
                           hideSectionMenu={hideSectionMenu}>
            {renderSections()}
        </StyledSectionMenu>
    );
};

SectionMenu.propTypes = {
    sections: PropTypes.array.isRequired,
    hideSectionMenu: PropTypes.bool
};
