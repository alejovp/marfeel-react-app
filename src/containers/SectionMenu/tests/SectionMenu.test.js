import React from 'react';
import { SectionMenu } from '../SectionMenu';
import theme from '../../../../__mocks__/theme.json';
import { mountWithTheme } from '../../../../scripts/jest/setup';
import { StyledSectionMenu } from '../StyledSectionMenu';
import { sectionsMock } from '../../../../__mocks__/sections';
import { SectionTab } from '../../../components/SectionTab';
import { DispatchContext, StateContext } from '../../../contexts/AppStore';
import { SET_SECTION_MENU_HEIGHT } from '../../../contexts/constants';

const mockHeight = 68;
jest.mock('../../../hooks', () => ({
    useElemDimensions: jest.fn().mockImplementation(() => ([
        jest.fn(),
        {
            height: mockHeight
        }
    ]))
}));

describe('<SectionMenu />', () => {

    describe('shape', () => {
        const renderedComponent = mountWithTheme(
            <DispatchContext.Provider value={jest.fn()}>
                <StateContext.Provider value={{}}>
                    <SectionMenu sections={sectionsMock} />
                </StateContext.Provider>
            </DispatchContext.Provider>, theme);

        it('will render correctly', () => {
            expect(renderedComponent).toMatchSnapshot();
        });

        it('will render a StyledSectionMenu component', () => {
            expect(renderedComponent.find(StyledSectionMenu)).toHaveLength(1);
        });

        it('will render as many SectionTab as items in sections prop', () => {
            expect(renderedComponent.find(SectionTab)).toHaveLength(10);
        });

    });

    describe('props', () => {
        const mockDispatch = jest.fn();

        it('will dispatch an action on component mount with SectionMenu height', () => {
            const renderedComponent = mountWithTheme(
                <DispatchContext.Provider value={mockDispatch}>
                    <StateContext.Provider value={{}}>
                        <SectionMenu sections={sectionsMock} />
                    </StateContext.Provider>
                </DispatchContext.Provider>, theme);
            renderedComponent.update();
            expect(mockDispatch).toBeCalledWith({ type: SET_SECTION_MENU_HEIGHT, payload: mockHeight });
        });

    });

});
