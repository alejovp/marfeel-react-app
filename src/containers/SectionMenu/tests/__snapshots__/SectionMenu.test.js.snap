// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SectionMenu /> shape will render correctly 1`] = `
<SectionMenu
  sections={
    Array [
      Object {
        "id": "sec-1",
        "path": "/section-1",
        "text": "SECTION 1",
      },
      Object {
        "id": "sec-2",
        "path": "/section-2",
        "text": "SECTION 2",
      },
      Object {
        "id": "sec-3",
        "path": "/section-3",
        "text": "SECTION 3",
      },
      Object {
        "id": "sec-4",
        "path": "/section-4",
        "text": "SECTION 4",
      },
      Object {
        "id": "sec-5",
        "path": "/section-5",
        "text": "SECTION 5",
      },
      Object {
        "id": "sec-6",
        "path": "/section-6",
        "text": "SECTION 6",
      },
      Object {
        "id": "sec-7",
        "path": "/section-7",
        "text": "SECTION 7",
      },
      Object {
        "id": "sec-8",
        "path": "/section-8",
        "text": "SECTION 8",
      },
      Object {
        "id": "sec-9",
        "path": "/section-9",
        "text": "SECTION 9",
      },
      Object {
        "id": "sec-10",
        "path": "/section-10",
        "text": "SECTION 10",
      },
    ]
  }
>
  <StyledSectionMenu
    dimensions={
      Object {
        "height": 68,
      }
    }
  >
    <ul
      className="gCdgFZ"
    >
      <SectionTab
        currentSection="sec-1"
        key="sec-1"
        onClick={[Function]}
        section={
          Object {
            "id": "sec-1",
            "path": "/section-1",
            "text": "SECTION 1",
          }
        }
      >
        <StyledSectionTab
          isActive={true}
          onClick={[Function]}
        >
          <li
            className="fxPcfg"
            onClick={[Function]}
          >
            <a>
              SECTION 1
            </a>
          </li>
        </StyledSectionTab>
      </SectionTab>
      <SectionTab
        currentSection="sec-1"
        key="sec-2"
        onClick={[Function]}
        section={
          Object {
            "id": "sec-2",
            "path": "/section-2",
            "text": "SECTION 2",
          }
        }
      >
        <StyledSectionTab
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="Oifry"
            onClick={[Function]}
          >
            <a>
              SECTION 2
            </a>
          </li>
        </StyledSectionTab>
      </SectionTab>
      <SectionTab
        currentSection="sec-1"
        key="sec-3"
        onClick={[Function]}
        section={
          Object {
            "id": "sec-3",
            "path": "/section-3",
            "text": "SECTION 3",
          }
        }
      >
        <StyledSectionTab
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="Oifry"
            onClick={[Function]}
          >
            <a>
              SECTION 3
            </a>
          </li>
        </StyledSectionTab>
      </SectionTab>
      <SectionTab
        currentSection="sec-1"
        key="sec-4"
        onClick={[Function]}
        section={
          Object {
            "id": "sec-4",
            "path": "/section-4",
            "text": "SECTION 4",
          }
        }
      >
        <StyledSectionTab
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="Oifry"
            onClick={[Function]}
          >
            <a>
              SECTION 4
            </a>
          </li>
        </StyledSectionTab>
      </SectionTab>
      <SectionTab
        currentSection="sec-1"
        key="sec-5"
        onClick={[Function]}
        section={
          Object {
            "id": "sec-5",
            "path": "/section-5",
            "text": "SECTION 5",
          }
        }
      >
        <StyledSectionTab
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="Oifry"
            onClick={[Function]}
          >
            <a>
              SECTION 5
            </a>
          </li>
        </StyledSectionTab>
      </SectionTab>
      <SectionTab
        currentSection="sec-1"
        key="sec-6"
        onClick={[Function]}
        section={
          Object {
            "id": "sec-6",
            "path": "/section-6",
            "text": "SECTION 6",
          }
        }
      >
        <StyledSectionTab
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="Oifry"
            onClick={[Function]}
          >
            <a>
              SECTION 6
            </a>
          </li>
        </StyledSectionTab>
      </SectionTab>
      <SectionTab
        currentSection="sec-1"
        key="sec-7"
        onClick={[Function]}
        section={
          Object {
            "id": "sec-7",
            "path": "/section-7",
            "text": "SECTION 7",
          }
        }
      >
        <StyledSectionTab
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="Oifry"
            onClick={[Function]}
          >
            <a>
              SECTION 7
            </a>
          </li>
        </StyledSectionTab>
      </SectionTab>
      <SectionTab
        currentSection="sec-1"
        key="sec-8"
        onClick={[Function]}
        section={
          Object {
            "id": "sec-8",
            "path": "/section-8",
            "text": "SECTION 8",
          }
        }
      >
        <StyledSectionTab
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="Oifry"
            onClick={[Function]}
          >
            <a>
              SECTION 8
            </a>
          </li>
        </StyledSectionTab>
      </SectionTab>
      <SectionTab
        currentSection="sec-1"
        key="sec-9"
        onClick={[Function]}
        section={
          Object {
            "id": "sec-9",
            "path": "/section-9",
            "text": "SECTION 9",
          }
        }
      >
        <StyledSectionTab
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="Oifry"
            onClick={[Function]}
          >
            <a>
              SECTION 9
            </a>
          </li>
        </StyledSectionTab>
      </SectionTab>
      <SectionTab
        currentSection="sec-1"
        key="sec-10"
        onClick={[Function]}
        section={
          Object {
            "id": "sec-10",
            "path": "/section-10",
            "text": "SECTION 10",
          }
        }
      >
        <StyledSectionTab
          isActive={false}
          onClick={[Function]}
        >
          <li
            className="Oifry"
            onClick={[Function]}
          >
            <a>
              SECTION 10
            </a>
          </li>
        </StyledSectionTab>
      </SectionTab>
    </ul>
  </StyledSectionMenu>
</SectionMenu>
`;
